lib/dal/src/workspace_snapshot/graph/v4/attribute_prototype/argument.rs

```diff
@@ -0,0 +1,100 @@
+use petgraph::{
+    prelude::*,
+    stable_graph::EdgeReference,
+};
+use si_id::AttributePrototypeArgumentId;
+
+use crate::{
+    EdgeWeight,
+    EdgeWeightKind,
+    EdgeWeightKindDiscriminants,
+    attribute::{
+        prototype::argument::value_source::ValueSource,
+        value::subscription::ValueSubscription,
+    },
+    workspace_snapshot::{
+        content_address::ContentAddress,
+        graph::{
+            WorkspaceSnapshotGraphError,
+            WorkspaceSnapshotGraphResult,
+            WorkspaceSnapshotGraphV4,
+            traits::attribute_prototype::argument::AttributePrototypeArgumentExt,
+        },
+        node_weight::{
+            NodeWeight,
+            traits::SiNodeWeight,
+        },
+    },
+};
+
+impl AttributePrototypeArgumentExt for WorkspaceSnapshotGraphV4 {
+    fn value_source_opt(
+        &self,
+        apa_id: AttributePrototypeArgumentId,
+    ) -> WorkspaceSnapshotGraphResult<Option<ValueSource>> {
+        let apa_index = self.get_node_index_by_id(apa_id)?;
```

Comment:
I'm curious. Why implement this as trait extension instead of as a method on AttributePrototypeArgument ?

Reply:
The rubric in my head is, common things that make multiple graph moves and/or need detailed graph knowledge, should live in graph land where it can be centralized and written a little more easily. Right now, extension traits are the blessed way to move something there. I'm not married to that rubric :)

Reply:
Moved out of traits for now during the split-graph transition period!